### PR TITLE
Initial WorkQueueService implementation.

### PR DESCRIPTION
**Pull request by klausw**
This limits execution to 4 parallel requests, adding one additional
parallel request for a currently-selected widget to prioritize that.

Includes a refactor to make status display conditional using CSS classes,
this reduces the amount of ng-show dependent code in gviz-directive.html.

Also disable animations for gviz chart elements, these were doing horrific
things to logic due to "animating" adding/removing classes so that
states ended up in a weird superposition of states that were meant to be
mutually exclusive.

### CODE REVIEW

**Comment by klausw** [comment]
This is still a work in progress, specifically missing docstrings and test coverage. Just wanted to get it out for feedback about the general approach, I don't feel I grok Angular sufficiently to be confident this is the right way to do it.

If we want to make the feature opt-in while preserving current logic, that will get more complex with some ugly conditionals in gviz-directive.js and elsewhere. Setting the default limit to a very large number would be easier.

If you want, I could pull out the "pk-if" CSS conditionals and disabling animations into a separate change since it would also be able to apply for the current logic.

**Comment by jmuharsky** [comment]
I don't think we want to make the feature opt-in.  We're trying to control quota limits for all users.  That said, I think the number of concurrent requests should be configurable (at least at the service level), and I'm good with setting the limit to a very large number.  If we really wanted "opt-out" behavior, though I don't think we do at present, we could just set it to the number of widgets on the dashboard.

As far as the animation fixes, I'd prefer those in a separate PR.  I'll take animation fixes anywhere I can get them. :)

File: client/components/explorer/explorer-header-directive.html [hunk 1]
```diff
@@ -1,6 +1,9 @@
 <div class="pk-header">
 
   <div class="pk-header-toolbar">
+    <span class="pk-header-status" ng-show="ctrl.workQueueSvc.canShowStatus()">
+      {{ ctrl.workQueueSvc.getStatus() }}
```
Comment (jmuharsky) [inline client/components/explorer/explorer-header-directive.html]: Angular-templated sections can get evaluated a lot; consider storing status as a property and updating the relevant strings on the appropriate events/methods, rather than calculating in the method.

File: client/components/util/work-queue-service.js [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+  this.queuePending_ = [];
+  this.queueExecuting_ = 0;
+};
+const WorkQueueService = explorer.components.util.WorkQueueService;
+
+WorkQueueService.NOTIFICATION = {
+  STARTED: 'Started'
+};
+
+WorkQueueService.MAX_PARALLEL_QUERIES_DEFAULT = 4;
```
Comment (jmuharsky) [inline client/components/util/work-queue-service.js]: This should be a service level setting, stored in [DataStore](https://github.com/GoogleCloudPlatform/PerfKitExplorer/blob/master/server/perfkit/explorer/model/explorer_config.py#L45), and exposed in [ConfigService](https://github.com/GoogleCloudPlatform/PerfKitExplorer/blob/master/client/components/config/config-service.js#L44) and the [config](https://github.com/GoogleCloudPlatform/PerfKitExplorer/blob/master/client/components/config/config-directive.html) dialog directive.

File: client/components/widget/data_viz/gviz/gviz-directive.js [hunk 3]
```diff
@@ -271,8 +254,9 @@ explorer.components.widget.data_viz.gviz.gvizChart = function(
           let promise = queryResultDataService.
               fetchResults(scope.widgetConfig);
           $timeout(function() {
-            scope.widgetConfig.state().datasource.status =
-                ResultsDataStatus.FETCHING;
+            // TODO(klausw): make behavior conditional on if queuing is enabled.
```
Comment (jmuharsky) [inline client/components/widget/data_viz/gviz/gviz-directive.js]: As we're not going to allow opt-out on this, this section seems safely removeable.

File: client/components/widget/data_viz/gviz/gviz-directive.js [hunk 4]
```diff
@@ -301,6 +285,15 @@ explorer.components.widget.data_viz.gviz.gvizChart = function(
               scope.widgetConfig.queryError = response.error;
             });
           });
+          // Progress notification
+          promise.then(null, null, angular.bind(this, function(notification) {
+            if (notification == 'Started' /*WorkQueueService.NOTIFICATION.STARTED*/) {
```
Comment (jmuharsky) [inline client/components/widget/data_viz/gviz/gviz-directive.js]: Add a workQueueService reference to the constructor to reference your constant.

File: client/components/widget/query/query-result-data-service.js [hunk 5]
```diff
@@ -27,6 +27,8 @@ goog.require('p3rf.perfkit.explorer.components.error.ErrorTypes');
 goog.require('p3rf.perfkit.explorer.components.error.ErrorService');
 goog.require('p3rf.perfkit.explorer.components.explorer.ExplorerService');
 goog.require('p3rf.perfkit.explorer.components.explorer.ExplorerStateService');
+goog.require('p3rf.perfkit.explorer.components.util.WorkQueueService');
+goog.require('p3rf.perfkit.explorer.models.ResultsDataStatus');
```
Comment (jmuharsky) [inline client/components/widget/query/query-result-data-service.js]: ResultsDataStatus isn't used in this file.

**Comment by jmuharsky** [comment]
Overall, it looks solid, though in addition to comments and tests, I'd prefer the CSS/style/animation changes to be a separate PR.

**Comment by klausw** [comment]
According to https://cloud.google.com/bigquery/quota-policy there's a limit of 50 concurrent interactive requests, and only 6 concurrent requests involving user-defined functions. So I think we definitely want a limit <= 50, and preferably a default limit that's significantly lower - maybe 10 or so?

If people start using UDFs, the limit would need to be far lower at the default quota limit - not sure if we can raise that and/or detect use of UDFs. If we can't do either, we may need to discourage using them.

I'll work on splitting out the animation and CSS changes into a separate PR, should be straighforward.

**Comment by jmuharsky** [comment]
To be clear here, I'm good with setting the user-defined limit to 5, or some global number (the 50 concurrent requests applies across the project, not just for a single user).  But for service administrators, I'm generally not a favor of hard boundaries, rather having a link on the setting that takes them to the docs for the limits.  That way, if/when BQ raises their concurrency (or a project moves to slotted execution, where concurrency is allowed at the expense of run time), admins can adjust accordingly.

File: client/components/widget/data_viz/gviz/gviz-directive.html [hunk 6]
```diff
@@ -1,14 +1,20 @@
...
-       class="pk-chart-loading md-primary md-hue-3"
-       ng-show="isLoadingDisplayed()">
-    <md-progress-circular md-mode="indeterminate"></md-progress-circular>
-  </div>
-  <div class="pk-chart-error"
-       ng-show="isErrorDisplayed()">
-    {{ widgetConfig.state().chart.error }}
+  <div class="pk-conditional" ng-class="getWidgetStatusClass()">
+    <div class="pk-if pk-cond-nodata pk-chart-nodata">
+      No data available.
```
Comment (jmuharsky) [inline client/components/widget/data_viz/gviz/gviz-directive.html]: This div needs styled to be similar to the chart error styling (centered, italic, greyed, etc.).

File: client/components/widget/data_viz/gviz/gviz-directive.js (hunk 4 above)
Comment (klausw) [inline client/components/widget/data_viz/gviz/gviz-directive.js]: Done via goog.require - it's a module-level constant and shouldn't need to be injected.

File: client/components/widget/data_viz/gviz/gviz-directive.js (hunk 3 above)
Comment (klausw) [inline client/components/widget/data_viz/gviz/gviz-directive.js]: Done.

File: client/components/widget/query/query-result-data-service.js (hunk 5 above)
Comment (klausw) [inline client/components/widget/query/query-result-data-service.js]: Removed. I was using this earlier for debugging but it's no longer needed.

File: client/components/explorer/explorer-header-directive.html (hunk 1 above)
Comment (klausw) [inline client/components/explorer/explorer-header-directive.html]: Done.

File: client/components/util/work-queue-service.js (hunk 2 above)
Comment (klausw) [inline client/components/util/work-queue-service.js]: Done, I split a small refactor into a separate commit in this PR.

**Comment by klausw** [comment]
Added tests and type annotations. Compiler warnings weren't working, I have a followup to get those
enabled but unfortunately I've only been able to fix ~400 of the ~600 pre-existing warnings so far.

Open items:
- 'nodata' div styling
- status display styling

File: client/components/widget/data_viz/gviz/gviz-directive.html (hunk 6 above)
Comment (klausw) [inline client/components/widget/data_viz/gviz/gviz-directive.html]: Done.

**Comment by klausw** [comment]
PTAL, I've fixed the styling for "no data" and the loading status display.

**Comment by jmuharsky** [comment]
LGTM
